### PR TITLE
Refactor incremental spmd algos

### PR DESCRIPTION
**Pull request by ethanglaser**
## Description

_Add a comprehensive description of proposed changes_

_List associated issue number(s) if exist(s): #6 (for example)_

_Documentation PR (if needed): #1340 (for example)_

_Benchmarks PR (if needed): https://github.com/IntelPython/scikit-learn_bench/pull/155 (for example)_

---

PR should start as a draft, then move to ready for review state after CI is passed and all applicable checkboxes are closed.
This approach ensures that reviewers don't spend extra time asking for regular requirements.

You can remove a checkbox as not applicable only if it doesn't relate to this PR in any way.
For example, PR with docs update doesn't require checkboxes for performance while PR with any change in actual code should have checkboxes and justify how this code change is expected to affect performance (or justification should be self-evident).

Checklist to comply with **before moving PR from draft**:

**PR completeness and readability**

- [ ] I have reviewed my changes thoroughly before submitting this pull request.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have updated the documentation to reflect the changes or created a separate PR with update and provided its number in the description, if necessary.
- [ ] Git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/uxlfoundation/scikit-learn-intelex/blob/main/CONTRIBUTING.md#pull-requests) for details)_.
- [ ] I have added a respective label(s) to PR if I have a permission for that.
- [ ] I have resolved any merge conflicts that might occur with the base branch.

**Testing**

- [ ] I have run it locally and tested the changes extensively.
- [ ] All CI jobs are green or I have provided justification why they aren't.
- [ ] I have extended testing suite if new functionality was introduced in this PR.

**Performance**

- [ ] I have measured performance for affected algorithms using [scikit-learn_bench](https://github.com/IntelPython/scikit-learn_bench) and provided at least summary table with measured data, if performance change is expected.
- [ ] I have provided justification why performance has changed or why changes are not expected.
- [ ] I have provided justification why quality metrics have changed or why changes are not expected.
- [ ] I have extended benchmarking suite and provided corresponding scikit-learn_bench PR if new measurable functionality was introduced in this PR.

### CODE REVIEW

File: onedal/spmd/linear_model/incremental_linear_model.py [hunk 1]
```diff
@@ -14,84 +14,14 @@
...
-        -------
-        self : object
-            Returns the instance itself.
-        """
-        module = super(base_IncrementalLinearRegression, self)._get_backend(
-            "linear_model", "regression"
-        )
-
-        self._queue = queue
-        policy = super(base_IncrementalLinearRegression, self)._get_policy(queue, X)
```
Comment (icfaust) [inline onedal/spmd/linear_model/incremental_linear_model.py:68]: I think this was done for the wonkiness of the get_backend  + when the spmd policy is used and when the dpc policy is used.

File: onedal/spmd/linear_model/incremental_linear_model.py [hunk 2]
```diff
@@ -14,84 +14,14 @@
...
-
-    API is the same as for `onedal.linear_model.IncrementalLinearRegression`.
-    """
-
-    def _reset(self):
-        self._partial_result = super(base_IncrementalLinearRegression, self)._get_backend(
-            "linear_model", "regression", "partial_train_result"
-        )
-
+    @support_input_format()
```
Comment (icfaust) [inline onedal/spmd/linear_model/incremental_linear_model.py:25]: Zero copy support will move away from support input format, as we will fully generate #2206 #2207 and #2189

**Review by icfaust — COMMENTED** [review]
I know this is still a WIP, but maybe we should have a chat about strategy?

File: onedal/spmd/linear_model/incremental_linear_model.py (hunk 1 above)
Comment (ethanglaser) [reply onedal/spmd/linear_model/incremental_linear_model.py:68]: Just significantly revised this

File: onedal/basic_statistics/incremental_basic_statistics.py [hunk 3]
```diff
@@ -71,8 +71,8 @@ def __init__(self, result_options="all"):
 
     def _reset(self):
         self._need_to_finalize = False
-        self._partial_result = self._get_backend(
-            "basic_statistics", None, "partial_compute_result"
+        self._partial_result = IncrementalBasicStatistics._get_backend(
```
Comment (ethanglaser) [inline onedal/basic_statistics/incremental_basic_statistics.py:74]: TODO: add comments

File: onedal/basic_statistics/incremental_basic_statistics.py [hunk 4]
```diff
@@ -71,8 +71,9 @@ def __init__(self, result_options="all"):
 
     def _reset(self):
         self._need_to_finalize = False
-        self._partial_result = self._get_backend(
-            "basic_statistics", None, "partial_compute_result"
+        # Not supported with spmd policy so IncrementalBasicStatistics must be specified
```
Comment (icfaust) [inline onedal/basic_statistics/incremental_basic_statistics.py:74]: Love the comment

File: onedal/basic_statistics/incremental_basic_statistics.py [hunk 5]
```diff
@@ -71,8 +71,9 @@ def __init__(self, result_options="all"):
...
+        self._partial_result = IncrementalBasicStatistics._get_backend(
+            IncrementalBasicStatistics, "basic_statistics", None, "partial_compute_result"
```
Comment (icfaust) [inline onedal/basic_statistics/incremental_basic_statistics.py:76]: @ahuber21 these changes are likely going to interact with #2168 (just as a heads up).

File: onedal/decomposition/pca.py [hunk 6]
```diff
@@ -128,15 +129,20 @@ def _create_model(self):
...
-        result = self._get_backend(
-            "decomposition", "dim_reduction", "infer", policy, params, model, X_table
+        # Not supported with spmd policy so BasePCA must be specified
+        result = BasePCA._get_backend(
+            BasePCA, "decomposition", "dim_reduction", "infer", policy, params, model, X_table
         )
         return from_table(result.transformed_data)
+    
+    def transform(self, X, queue=None):
+        return self.predict(X, queue=queue)
```
Comment (icfaust) [inline onedal/decomposition/pca.py:145]:
```suggestion
    transform = predict
```
unless you need to have something specific with ```IncrementalPCA.transform.__name__``` that we need to worry about

File: onedal/decomposition/tests/test_incremental_pca.py [hunk 7]
```diff
@@ -40,7 +40,7 @@ def test_on_gold_data(queue, is_deterministic, whiten, num_blocks, dtype):
 
     result = incpca.finalize_fit()
 
-    transformed_data = incpca.predict(X, queue=queue)
+    transformed_data = incpca.transform(X, queue=queue)
```
Comment (icfaust) [inline onedal/decomposition/tests/test_incremental_pca.py:43]: Just for conformance purposes to sklearn (though not strictly necessary in the onedal folder)?

**Review by icfaust — COMMENTED** [review]
I'll be honest, the get_backend stuff just seems to be a headache through and through. Small stuff but overall really good work.

File: onedal/decomposition/pca.py (hunk 6 above)
Comment (ethanglaser) [reply onedal/decomposition/pca.py:145]: good call

File: onedal/decomposition/tests/test_incremental_pca.py (hunk 7 above)
Comment (ethanglaser) [reply onedal/decomposition/tests/test_incremental_pca.py:43]: exactly - pca predict does not exist in sklearn so would prefer to not use this convention if possible

**Comment by ethanglaser** [comment]
Would like to merge this before https://github.com/uxlfoundation/scikit-learn-intelex/pull/2153
